File: BOJ/Silver/14888. 연산자 끼워넣기/Main.py
# import sys
# sys.stdin = open("input.txt", "r")

# 백준 파이썬 14888번 연산자 끼워넣기

# 입력
N = int(input())
arr = list(map(int, input().split()))
oper0 = list(map(int, input().split()))

# 변수 선언
max_result = 0
min_result = float('inf')

def cal(oper_num, now, result, oper):
    global max_result, min_result
    
    # 종료 조건: 연산자(oper_num)를 다 썼을 때
    # 최대, 최소 값 저장
    if oper_num == 0:
        max_result = max(max_result, result)
        min_result = min(min_result, result)
        return

    # 들어갈 때: oper_num -1
    for i in range(len(oper)):
        if oper[i]:
            if i == 0:
                # operator = '+'
                oper[i] -= 1
                cal(oper_num-1, now+1, result + arr[now+1], oper)
                oper[i] += 1 
            elif i == 1:
                # operator = '-'
                oper[i] -= 1
                cal(oper_num-1, now+1, result - arr[now+1], oper)
                oper[i] += 1 
            elif i == 2:
                # operator = '*'
                oper[i] -= 1
                cal(oper_num-1, now+1, result * arr[now+1], oper)
                oper[i] += 1 
            else:
                # operator = '/'
                oper[i] -= 1
                if result <0:
                    cal(oper_num-1, now+1, -(-result // arr[now+1]), oper)
                else:
                    cal(oper_num-1, now+1, result // arr[now+1], oper)
                oper[i] += 1 

cal(N-1, 0, arr[0], oper0)

print(max_result)
print(min_result)

File: BOJ/Silver/14888. 연산자 끼워넣기/Main.cc
#define _CRT_SECURE_NO_WARNINGS
#include <cstdio>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int min_result, max_result;

void cal(int now, vector<int>& card, vector<int>& arr, int result) {
	// 종료 조건: 더 이상 연산할 연산자 없을 경우 종료
	if (now == arr.size() - 1) {
		max_result = max(max_result, result);
		min_result = min(min_result, result);
		return;
	}

	for (int i = 0; i < 4; i++) {
		if (card[i] > 0) {
			card[i] --;

			// + 인 경우
			if (i == 0) 
				cal(now + 1, card, arr, result + arr[now+1]);

			// - 인 경우
			else if (i == 1) 
				cal(now + 1, card, arr, result - arr[now + 1]);

			// * 인 경우
			else if (i == 2) 
				cal(now + 1, card, arr, result * arr[now + 1]);

			// / 인 경우
            else if (i == 3) {
                if (result < 0)
                    cal(now + 1, card, arr, -(-result / arr[now + 1]));
                else
                    cal(now + 1, card, arr, result / arr[now + 1]);
            }
			card[i] ++;
		}
	}
	
}

int main() {

    // N: 숫자의 개수
    int N = 0;
    cin >> N;
    
    // arr: 수식에 들어갈 숫자
    vector<int>arr(N);
    for (int i = 0; i < N; i++) cin >> arr[i];

    // card: 연산자 카드의 개수
    vector<int>card(4);
    for (int i = 0; i < 4; i++) cin >> card[i];

    // 계산
    max_result = -1e9;
    min_result = 1e9;
    cal(0, card, arr, arr[0]);

    cout << max_result << '\n';
    cout << min_result;
	
}